### PR TITLE
[DSD][Origin Detection] Allow DSD clients to change tags cardinality

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

Dogstatsd clients can now choose the cardinality of tags added by origin detection via the tag `dd.internal.card` (low, orch, high).

### Motivation

Support a dynamic cardinality config

### Describe your test plan

`echo -n "dsd.testing.<CARD>:1|g|#foo:bar,dd.internal.entity_id:<Pod UID>,dd.internal.card:<CARD>" | nc -u -w1 <Agent IP> 8125`

`<CARD>` can be `low`, `orch` or `high`

### CODE REVIEW

**Review by L3n41c — APPROVED** [review]

File: pkg/dogstatsd/enrich.go [hunk 1]
```diff
@@ -48,7 +54,21 @@ func extractTagsMetadata(tags []string, defaultHostname string, originTags strin
...
+		return tagger.DogstatsdCardinality
+	}
+
+	taggerCardinality, err := collectors.StringToTagCardinality(cardinality)
+	if err != nil {
+		log.Tracef("Couldn't convert cardinality tag: %w", err)
+		return tagger.DogstatsdCardinality
+	}
+
+	return taggerCardinality
```
Comment (hush-hush) [inline pkg/dogstatsd/enrich.go:71]: Performances are really tight in dogstatsd. Could we return a string from `extractTagsMetadata` and do this either in the tagger directly ?

it would also avoid the dependency to tagger/collectors everywhere.

File: pkg/dogstatsd/enrich.go [hunk 2]
```diff
@@ -69,9 +89,9 @@ func enrichMetricType(dogstatsdMetricType metricType) metrics.MetricType {
...
 }
 
-func isBlacklisted(metricName, namespace string, namespaceBlacklist []string) bool {
+func isExcluded(metricName, namespace string, namespaceExcluded []string) bool {
 	if namespace != "" {
-		for _, prefix := range namespaceBlacklist {
+		for _, prefix := range namespaceExcluded {
 			if strings.HasPrefix(metricName, prefix) {
 				return true
 			}
```
Comment (hush-hush) [inline pkg/dogstatsd/enrich.go:97]: Why the change in naming ?

File: releasenotes/notes/control-dsd-cardinality-via-tags-c0f8b08f424deeda.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    Dogstatsd clients can now choose the cardinality of tags added by origin detection
+    via the tag dd.internal.cardinality (low, orchestrator, high).
```
Comment (hush-hush) [inline releasenotes/notes/control-dsd-cardinality-via-tags-c0f8b08f424deeda.yaml:5]: WDYT ?

```suggestion
    Dogstatsd clients can now choose the cardinality of tags added by origin detection per metrics
    via the tag 'dd.internal.cardinality' (low, orchestrator, high).
```

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/dogstatsd/enrich.go (hunk 1 above)
Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go:71]: Actually `collectors.TagCardinality` is an `int`, sorry if I'm missing something obvious but could you elaborate why this can raise performance concerns?

In addition, regardless of the type, whether we run this in dsd or in the tagger it would have the same impact on dsd right? because dsd calls `tagger.TagBuilder` which is a blocking function. Not sure how moving this logic to the tagger could help.

In any case, the main reason why I've put this logic here is to avoid changing the tagger API which expects a `collectors.TagCardinality`

`TagBuilder(entity string, cardinality collectors.TagCardinality, tb *util.TagsBuilder) error`

File: pkg/dogstatsd/enrich.go (hunk 2 above)
Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go:97]: Will make the change in a separate PR 👍

Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go]: https://github.com/DataDog/datadog-agent/pull/7559

File: pkg/dogstatsd/enrich.go (hunk 1 above)
Comment (hush-hush) [reply pkg/dogstatsd/enrich.go:71]: This is the loop that process dogstatsd packet. The slower it is the bigger are the packets drop. That's why we try to push as much as the post-processing later in the pipeline as we can.

Dogstatsd actually don't call the tagger anymore, the aggregator does. We move the hashing later to reduce allocation and GC run time.

We could do this in `enrichTag` to not change the tagger API then.

WDYT ?

Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go:71]: Thanks for the explanation, I have a clearer picture now! Moving it to `enrichTags` sounds good 👍

Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go]: PR updated

File: pkg/metrics/metric_sample.go [hunk 4]
```diff
@@ -149,3 +150,19 @@ func (m *MetricSample) Copy() *MetricSample {
 	copy(dst.Tags, m.Tags)
 	return dst
 }
+
+// taggerCardinality converts tagger cardinality string to collectors.TagCardinality
+// It defaults to DogstatsdCardinality if the string is empty or unknown
+func taggerCardinality(cardinality string) collectors.TagCardinality {
```
Comment (olivielpeau) [inline pkg/metrics/metric_sample.go:156]: I know this has been discussed with @hush-hush, but I'm not sure doing the `string` to `int` (`collectors.TagCardinality`) conversion here instead of earlier in the pipeline will actually improve performance: doing the string conversion here means that the `MetricSample` has to keep a string instead of just an `int` value.

Since I believe this string is interned in the string interner, it doesn't put much additional pressure on the GC (I think, but not sure), so it may not have any noticeable impact, but worth discussing. Unfortunately I can't think of a good micro-benchmark to measure the impact.

In any case this shouldn't block the PR, but happy to hear your thoughts @hush-hush

**Review by olivielpeau — COMMENTED** [review]
Could you run before/after benchmarks from `pkg/dogstatsd/enrich_bench_test.go`, just as a sanity check?

I left a comment on the tag key/values to use, let me know your thoughts.

The other comment is a perf nit that shouldn't block the PR.

**Comment by ahmed-mez** [comment]
Thank you @olivielpeau for the review! Here is the benchmark results

Before

```
BenchmarkExtractTagsMetadata
BenchmarkExtractTagsMetadata/20-tags
BenchmarkExtractTagsMetadata/20-tags-8         	31696617	        35.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/40-tags
BenchmarkExtractTagsMetadata/40-tags-8         	25163206	        47.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/60-tags
BenchmarkExtractTagsMetadata/60-tags-8         	20807089	        64.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/80-tags
BenchmarkExtractTagsMetadata/80-tags-8         	13863537	        88.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/100-tags
BenchmarkExtractTagsMetadata/100-tags-8        	11995491	       102 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/120-tags
BenchmarkExtractTagsMetadata/120-tags-8        	10091775	       118 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/140-tags
BenchmarkExtractTagsMetadata/140-tags-8        	 9281592	       135 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/160-tags
BenchmarkExtractTagsMetadata/160-tags-8        	 8267480	       148 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/180-tags
BenchmarkExtractTagsMetadata/180-tags-8        	 7227134	       163 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/200-tags
BenchmarkExtractTagsMetadata/200-tags-8        	 6457825	       186 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/DataDog/datadog-agent/pkg/dogstatsd	13.340s
```

After
```
BenchmarkExtractTagsMetadata
BenchmarkExtractTagsMetadata/20-tags
BenchmarkExtractTagsMetadata/20-tags-8         	20539608	        63.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/40-tags
BenchmarkExtractTagsMetadata/40-tags-8         	25886548	        46.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/60-tags
BenchmarkExtractTagsMetadata/60-tags-8         	16819836	        71.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/80-tags
BenchmarkExtractTagsMetadata/80-tags-8         	14505698	        85.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/100-tags
BenchmarkExtractTagsMetadata/100-tags-8        	12084690	        97.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/120-tags
BenchmarkExtractTagsMetadata/120-tags-8        	10756652	       113 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/140-tags
BenchmarkExtractTagsMetadata/140-tags-8        	 9584504	       129 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/160-tags
BenchmarkExtractTagsMetadata/160-tags-8        	 8116441	       147 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/180-tags
BenchmarkExtractTagsMetadata/180-tags-8        	 7447725	       160 ns/op	       0 B/op	       0 allocs/op
BenchmarkExtractTagsMetadata/200-tags
BenchmarkExtractTagsMetadata/200-tags-8        	 6847184	       161 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/DataDog/datadog-agent/pkg/dogstatsd	15.994s
```

File: pkg/dogstatsd/enrich.go [hunk 5]
```diff
@@ -8,21 +8,24 @@ import (
 )
 
 var (
-	hostTagPrefix       = "host:"
-	entityIDTagPrefix   = "dd.internal.entity_id:"
-	entityIDIgnoreValue = "none"
+	hostTagPrefix        = "host:"
+	entityIDTagPrefix    = "dd.internal.entity_id:"
+	entityIDIgnoreValue  = "none"
+	cardinalityTagPrefix = "dd.internal.cardinality:"
```
[the remark this answers is not in the document]
Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go:14]: This makes sense yes. Especially if we plan to contribute to the official dsd clients, all of this would become hidden and users won't have to deal directly with the tag name and the keys/values.
I'll update the PR accordingly!

File: pkg/metrics/metric_sample.go (hunk 4 above)
Comment (hush-hush) [reply pkg/metrics/metric_sample.go:156]: The string should be in the interner. And doing it later in the pipeline allows us to use the channel between the aggregator and dogstatsd to absorb spike instead of holding a `packet` from the packet pool longer.

File: releasenotes/notes/control-dsd-cardinality-via-tags-c0f8b08f424deeda.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    Dogstatsd clients can now choose the cardinality of tags added by origin detection per metrics
+    via the tag 'dd.internal.card' (low, orchestrator, high).
```
Comment (hush-hush) [inline releasenotes/notes/control-dsd-cardinality-via-tags-c0f8b08f424deeda.yaml:5]:
```suggestion
    via the tag 'dd.internal.card' ("low", "orch", "high").
```

**Review by hush-hush — APPROVED** [review]
Looking good, added a quick comment for the release note.

File: pkg/metrics/metric_sample.go (hunk 4 above)
Comment (olivielpeau) [reply pkg/metrics/metric_sample.go:156]: ok, thanks for chiming in and confirming this string is interned. I think it's hard to predict the impact without an e2e benchmark though, but not a blocker now

Comment (ahmed-mez) [reply pkg/metrics/metric_sample.go:156]: thank you both! @olivielpeau unless you have other suggestions, could you add your +1 so we can merge the PR?

**Review by olivielpeau — APPROVED** [review]
